File: 50.powx-n/solution.py
class Solution:
    def __init__(self):
        self.cache = {}

    def myPow(self, x: float, n: int) -> float:
        if n == 1:
            return x
        if n == 0:
            return 1
        if n == -1:
            return 1 / x
        if n in self.cache:
            return self.cache[n]
        p = self.myPow(x, n // 2) * self.myPow(x, n - n // 2)
        self.cache[n] = p
        return p

File: 50.powx-n/solution.kt
class Solution {
    fun myPow(x: Double, n: Int): Double {
        if (n == 0) return 1.0
        val m = HashMap<Int, Double>()
        fun pow(n: Int): Double {
            if (n == -1) return 1/x
            if (n == 1) return x
            val vn = m[n]
            if (vn != null) return vn
            val v_1 = m[n-1]
            if (v_1 != null) return v_1 * x
            val v = pow(n/2) * pow(n-n/2)
            m[n] = v
            return v
        }
        return pow(n)
    }
}
